Library/Homebrew/utils/github.rb

```diff
@@ -1,301 +1,91 @@
-          #{CREATE_GITHUB_PAT_MESSAGE}
-        EOS
-      end
-      super message.freeze
+      odeprecated "GitHub::AuthenticationFailedError", "GitHub::API::AuthenticationFailedError"
+      super
     end
   end
 
-  # Error when the user has no GitHub API credentials set at all (macOS keychain or envvar).
-  class MissingAuthenticationError < Error
+  # Wrapper for GitHub::API::MissingAuthenticationError
+  class MissingAuthenticationError < API::MissingAuthenticationError
     def initialize
-      message = +"No GitHub credentials found in macOS Keychain or environment.\n"
-      message << CREATE_GITHUB_PAT_MESSAGE
-      super message
+      odeprecated "GitHub::MissingAuthenticationError", "GitHub::API::MissingAuthenticationError"
+      super
     end
   end
 
-  # Error when the API returns a validation error.
-  class ValidationFailedError < Error
+  # Wrapper for GitHub::API::ValidationFailedError
+  class ValidationFailedError < API::ValidationFailedError
     def initialize(github_message, errors)
-      @github_message = if errors.empty?
-        github_message
-      else
-        "#{github_message}: #{errors}"
-      end
-
-      super(@github_message)
+      odeprecated "GitHub::ValidationFailedError", "GitHub::API::ValidationFailedError"
+      super
     end
   end
 
-  API_ERRORS = [
-    AuthenticationFailedError,
-    HTTPNotFoundError,
-    RateLimitExceededError,
-    Error,
-    JSON::ParserError,
-  ].freeze
-
-  # Gets the password field from `git-credential-osxkeychain` for github.com,
-  # but only if that password looks like a GitHub Personal Access Token.
-  sig { returns(T.nilable(String)) }
   def keychain_username_password
-    github_credentials = Utils.popen(["git", "credential-osxkeychain", "get"], "w+") do |pipe|
-      pipe.write "protocol=https\nhost=github.com\n"
-      pipe.close_write
-      pipe.read
-    end
-    github_username = github_credentials[/username=(.+)/, 1]
-    github_password = github_credentials[/password=(.+)/, 1]
-    return unless github_username
-
-    # Don't use passwords from the keychain unless they look like
-    # GitHub Personal Access Tokens:
-    #   https://github.com/Homebrew/brew/issues/6862#issuecomment-572610344
-    return unless /^[a-f0-9]{40}$/i.match?(github_password)
-
-    github_password
-  rescue Errno::EPIPE
-    # The above invocation via `Utils.popen` can fail, causing the pipe to be
-    # prematurely closed (before we can write to it) and thus resulting in a
-    # broken pipe error. The root cause is usually a missing or malfunctioning
-    # `git-credential-osxkeychain` helper.
-    nil
+    odeprecated "GitHub.keychain_username_password", "GitHub::API.keychain_username_password"
+    API.keychain_username_password
   end
 
   def api_credentials
-    @api_credentials ||= begin
-      Homebrew::EnvConfig.github_api_token || keychain_username_password
-    end
+    odeprecated "GitHub.api_credentials", "GitHub::API.api_credentials"
```

Comment:
May need to deprecate these in the new minor release instead? What does @Homebrew/brew think? Also/instead: I think you can limit these to the only ones you've specifically seen used elsewhere for now?

Reply:
Since this is private API, we can forgo the usual deprecation process.

However, since this is used in other Homebrew repositories, we should consider whether these should actually be public API instead. My view is private API usage in other repositories should be kept to a minimum.

Reply:
I guess usage isn't widespread: I found 6 instances of `open_api` usage so far, outside of `Homebrew/brew`. These were spread across 3 of our repositories (`homebrew-linux-dev`, `actions` and `homebrew-cask`).

> Also/instead: I think you can limit these to the only ones you've specifically seen used elsewhere for now?

Alright, seems like it might only be `open_api`, but I'll check again to be sure.

Reply:
> Since this whole module is private API, we can forgo the usual deprecation process.

If it's specified/documented as such: I agree.

> However, since some of these are used in other Homebrew repositories, we should consider whether these should actually be public API instead. My view is private API usage in other repositories should be kept to a minimum.

Agreed but this isn't always feasible. Homebrew/homebrew-bundle, for example, makes extensive use of private APIs because it's too slow to do otherwise. We could make these public, of course.

> Alright, seems like it might only be `open_api`, but I'll check again to be sure.

Cool, in that case then: yeh, just a wrapper for that seems sufficient.

Reply:
> If it's specified/documented as such: I agree.

`@api private` is our documentation here. Labels it nicely here: https://rubydoc.brew.sh/GitHub.html

> Agreed but this isn't always feasible.

For sure.

My general thinking is: what are we doing in these repositories that we think that no one outside the Homebrew org should be replicating?